### PR TITLE
[FLINK-25288][tests] add savepoint and metric test cases in source suite of connector testframe

### PR DESCRIPTION
**Pull request by ruanhang1993**
## What is the purpose of the change

This pull request adds savepoint and metric test cases in source suite of the connector testfram.

## Brief change log 

  - Add savepoint and metric test cases in the source suite of the connector testfram
  - Change the tests in the Kafka and Pulsar connector

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by ruanhang1993** [comment]
@flinkbot run azure

**Comment by ruanhang1993** [comment]
@flinkbot run azure

**Comment by ruanhang1993** [comment]
@flinkbot run azure

**Comment by ruanhang1993** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/testutils/KafkaSourceExternalContext.java [hunk 1]
```diff
@@ -58,20 +58,20 @@
...
 
     private final List<URL> connectorJarPaths;
-    private final String bootstrapServers;
-    private final String topicName;
-    private final SplitMappingMode splitMappingMode;
-    private final AdminClient adminClient;
+    protected final String bootstrapServers;
+    protected final String topicName;
+    protected final SplitMappingMode splitMappingMode;
+    protected final AdminClient adminClient;
```
Comment (leonardBang) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/testutils/KafkaSourceExternalContext.java:71]: Do we need to change this privilege? I didn't find any child visit this.

File: flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/FlinkContainerTestEnvironment.java [hunk 2]
```diff
@@ -68,6 +71,7 @@ public FlinkContainerTestEnvironment(
         config.set(HEARTBEAT_INTERVAL, 1000L);
         config.set(HEARTBEAT_TIMEOUT, 5000L);
         config.set(SLOT_REQUEST_TIMEOUT, 10000L);
+        config.set(METRIC_FETCHER_UPDATE_INTERVAL, 1000L);
```
Comment (leonardBang) [inline flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/FlinkContainerTestEnvironment.java:74]: We can use constant to maintain these magic number, these are used for tests with default constant value IIUC.
btw, Could connector developer pass custom Configuration?

File: flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/AvroBulkFormatITCase.java [hunk 3]
```diff
@@ -64,6 +66,10 @@
 
     private static final RowDataSerializer SERIALIZER = new RowDataSerializer(ROW_TYPE);
 
+    @SuppressWarnings("unused")
```
Comment (leonardBang) [inline flink-formats/flink-avro/src/test/java/org/apache/flink/formats/avro/AvroBulkFormatITCase.java:69]: hint: add `@SuppressWarnings("unused")` for all unused annotation fields

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/environment/MiniClusterTestEnvironment.java [hunk 4]
```diff
@@ -53,9 +61,12 @@
     private boolean isStarted = false;
 
     public MiniClusterTestEnvironment() {
+        Configuration conf = new Configuration();
+        conf.set(METRIC_FETCHER_UPDATE_INTERVAL, 1000L);
```
Comment (leonardBang) [inline flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/environment/MiniClusterTestEnvironment.java:65]: minor: Use constant for specific test configuration value.

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/environment/MiniClusterTestEnvironment.java [hunk 5]
```diff
@@ -71,7 +82,18 @@ public MiniClusterTestEnvironment() {
     @Override
     public StreamExecutionEnvironment createExecutionEnvironment(
             TestEnvironmentSettings envOptions) {
-        return StreamExecutionEnvironment.getExecutionEnvironment();
+        Configuration configuration = new Configuration();
+        if (envOptions.getSavepointRestorePath() != null) {
+            configuration.set(SAVEPOINT_PATH, envOptions.getSavepointRestorePath());
```
Comment (leonardBang) [inline flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/environment/MiniClusterTestEnvironment.java:87]:
```suggestion
            configuration.setString(SAVEPOINT_PATH, envOptions.getSavepointRestorePath());
```

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/junit/annotations/Semantic.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Marks the field in test class defining supported semantic: {@link
+ * org.apache.flink.streaming.api.CheckpointingMode}.
+ *
+ * <p>Only one field can be annotated in test class.
+ */
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@Experimental
+public @interface Semantic {}
```
Comment (leonardBang) [inline flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/junit/annotations/Semantic.java:37]:
```suggestion
public @interface TestSemantics {}
```

We can use this in following way IIUC, right ? 
```
@TestSemantics
    CheckpointingMode[] testSemantics = new CheckpointingMode[] {CheckpointingMode.EXACTLY_ONCE, CheckpointingMode.AT_LEAST_ONCE};
```
```

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SourceTestSuiteBase.java [hunk 7]
```diff
@@ -86,6 +103,8 @@
 public abstract class SourceTestSuiteBase<T> {
 
     private static final Logger LOG = LoggerFactory.getLogger(SourceTestSuiteBase.class);
+    static ExecutorService executorService =
+            Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors() * 2);
```
Comment (leonardBang) [inline flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SourceTestSuiteBase.java:107]: `Runtime.getRuntime().availableProcessors() * 2` is too expensive from my side, and could you add a `private` decoration too.

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/MetricQueryer.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.util.Iterator;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+/** The queryer is used to get job metrics by rest API. */
```
Comment (leonardBang) [inline flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/MetricQueryer.java:50]:
```suggestion
/** The querier  used to get job metrics by rest API. */
```

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/MetricQueryer.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+                .sendRequest(
+                        endpoint.getAddress(),
+                        endpoint.getPort(),
+                        AggregatedSubtaskMetricsHeaders.getInstance(),
+                        subtaskMetricsParameters,
+                        EmptyRequestBody.getInstance())
+                .get(30, TimeUnit.SECONDS);
+    }
+
+    public Double getMetricByRestApi(
```
Comment (leonardBang) [inline flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/MetricQueryer.java:120]: Maybe `getAggregatedMetricsByRestAPI` is more accurate as the function returns `Double` type value.

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/TestDataMatchers.java [hunk 10]
```diff
@@ -33,180 +36,194 @@
...
-            }
-        }
-
-        @Override
-        protected boolean matchesSafely(Iterator<T> resultIterator, Description description) {
-            if (mismatchDescription != null) {
-                description.appendText(mismatchDescription);
-                return false;
-            }
+        protected boolean matchAtLeastOnce(Iterator<T> resultIterator) {
```
Comment (leonardBang) [inline flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/TestDataMatchers.java:174]: Add unit test for `AtLeastOnce` case ?

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/TestUtils.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+            return result;
+        } else if (AT_LEAST_ONCE.equals(semantic)) {
+            while (retryIndex++ < retryTimes && !containSameVal(expected, result, semantic)) {
+                result.addAll(reader.poll(Duration.ofMillis(timeoutMs)));
+            }
+            return result;
+        }
+        throw new IllegalStateException(
+                String.format("%s delivery guarantee doesn't support test.", semantic.name()));
+    }
```
Comment (leonardBang) [inline flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/TestUtils.java:73]: These methods never used ?

File: flink-test-utils-parent/flink-connector-test-utils/src/test/java/org/apache/flink/connector/testframe/utils/TestDataMatchersTest.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+/** Unit test for {@link TestDataMatchers}. */
+public class TestDataMatchersTest {
```
Comment (leonardBang) [inline flink-test-utils-parent/flink-connector-test-utils/src/test/java/org/apache/flink/connector/testframe/utils/TestDataMatchersTest.java:41]: Add tests for `CheckpointingMode.AT_LEAST_ONCE`

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java [hunk 13]
```diff
@@ -40,6 +42,9 @@
     @TestExternalSystem
     PulsarTestEnvironment pulsar = new PulsarTestEnvironment(PulsarRuntime.mock());
 
+    @Semantic
+    CheckpointingMode[] semantics = new CheckpointingMode[] {CheckpointingMode.EXACTLY_ONCE};
```
Comment (leonardBang) [inline flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java:46]: I didn't find any ITCase for `CheckpointingMode.AT_LEAST_ONCE`,  we need add some case as we introduced the semantics  supports.

**Review by leonardBang — COMMENTED** [review]
Thanks @ruanhang1993 for the contribution, I left some comments

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/junit/annotations/Semantic.java (hunk 6 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/junit/annotations/Semantic.java:37]: Yes, we could use like this.
The changes about the semantic tests are in the PR #18547. I think this change could be applied in that PR.

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/TestDataMatchers.java (hunk 10 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/TestDataMatchers.java:174]: This has been fixed in PR #18547 .

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/TestUtils.java (hunk 11 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/TestUtils.java:73]: Yes, I will remove these useless codes.

File: flink-test-utils-parent/flink-connector-test-utils/src/test/java/org/apache/flink/connector/testframe/utils/TestDataMatchersTest.java (hunk 12 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-connector-test-utils/src/test/java/org/apache/flink/connector/testframe/utils/TestDataMatchersTest.java:41]: This has been fixed in PR #18547 .

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SourceTestSuiteBase.java (hunk 7 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/testsuites/SourceTestSuiteBase.java:107]: I will remove this field in this class.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/testutils/KafkaSourceExternalContext.java (hunk 1 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/testutils/KafkaSourceExternalContext.java]: This is invoked in another PR. I will revert this.

File: flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/FlinkContainerTestEnvironment.java (hunk 2 above)
Comment (ruanhang1993) [reply flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/FlinkContainerTestEnvironment.java]: fixed

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/environment/MiniClusterTestEnvironment.java (hunk 4 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/environment/MiniClusterTestEnvironment.java]: fixed

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/environment/MiniClusterTestEnvironment.java (hunk 5 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/environment/MiniClusterTestEnvironment.java]: fixed

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/MetricQueryer.java (hunk 8 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/MetricQueryer.java]: fixed

File: flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/MetricQueryer.java (hunk 9 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/MetricQueryer.java]: fixed

File: flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java (hunk 13 above)
Comment (ruanhang1993) [reply flink-connectors/flink-connector-pulsar/src/test/java/org/apache/flink/connector/pulsar/source/PulsarSourceITCase.java:46]: Actually, this semantic tests are contained in the sink tests for Kafka connector. There are not ITCase in the source now.

**Review by leonardBang — APPROVED** [review]
Thanks @ruanhang1993 for the contribution, LGTM. I also append one commit for minor improvement.

**Pull request closed without merging**